llama_index/llms/openai_like.py

```diff
@@ -0,0 +1,54 @@
+from typing import Any, Optional
+
+from llama_index.bridge.pydantic import Field
+from llama_index.constants import DEFAULT_CONTEXT_WINDOW
+from llama_index.llms import OpenAI
+
+
+class OpenAILike(OpenAI):
+    """
+    OpenAILike is a thin wrapper around the OpenAI model that makes it compatible with 3rd party tools
+    that provide an openai-compatible api
+
+    Currently, llama_index prevents using custom models with their OpenAI class because they need to be able to infer some metadata from the model name.
+
+    NOTE: You still need to set the OPENAI_BASE_API and OPENAI_KEY environment variables. OPENAI_KEY can normally be set to anything in this case, but will depend on the tool you're using.
+    """
+
+    context_window: int = Field(
+        default=DEFAULT_CONTEXT_WINDOW,
+        description="The maximum number of context tokens for the model.",
+    )
+    is_chat_model: bool = Field(
+        default=False, description="Indicates that the custom model is a chat_model."
+    )
+    is_function_calling_model: bool = Field(
+        default=False,
+        description="Indicates that the custom model is a function calling model.",
+    )
+    tokenizer: Optional[str] = Field(
+        default=None,
+        description="The name of a huggingface transformers tokenizer to use for this model",
```

Comment:
Tokenizers can come from other places than Hugging Face, I would not specify the source.  Just say a tokenizer

Reply:
See [this comment](https://github.com/run-llama/llama_index/pull/7973#discussion_r1349761780).  Really you just need an object that has an `encode` function.

Check how `HuggingFaceLLM` handles this with being able to pass a `model_name` or a `model`: https://github.com/run-llama/llama_index/blob/v0.8.41/llama_index/llms/huggingface.py#L91-L92

Note that that `HuggingFaceLLM` handles it with two separate args is suboptimal, because it violates the "one obvious path" principle. 

Imo this should be the same: able to pass:
- A tokenizer model with an `encode` function (it can come from anywhere)
- A name of a tokenizer model on Hugging Face, for use with `transformers. AutoTokenizer`

The way I would do it is have the type be `Union[_Tokenizer, str]`:
- `_Tokenizer`: can be passed a model
- `str`: this is the name used with `transformers. AutoTokenizer`

Then the `description` would be `"A tokenizer model instance, or the name of a Hugging Face tokenizer model (requires transformers package)"`